subsys/bluetooth/audio/has_client.c

```diff
@@ -0,0 +1,770 @@
+}
+
+static uint8_t active_preset_idx_ntf_cb(struct bt_conn *conn,
+					struct bt_gatt_subscribe_params *params, const void *data,
+					uint16_t length)
+{
+	struct bt_has_client *client = client_by_conn(conn);
+	struct net_buf_simple buf;
+	uint8_t id;
+
+	BT_DBG("conn %p len %u", conn, length);
+
+	if (length == 0) {
+		return BT_GATT_ITER_CONTINUE;
+	}
+
+	net_buf_simple_init_with_data(&buf, (void *)data, length);
+
+	id = net_buf_simple_pull_u8(&buf);
+
+	if (id != client->has.active_id) {
+		client->has.active_id = id;
+		has_cb->active_preset(&client->has, 0, client->has.active_id);
+	}
+
+	return BT_GATT_ITER_CONTINUE;
+}
+
+static void cfg_active_preset_idx_ntf_cb(struct bt_conn *conn, uint8_t err,
+					 struct bt_gatt_write_params *write)
+{
+	struct bt_has_client *client = client_by_conn(conn);
+
+	BT_DBG("conn %p err 0x%02x", client->conn, err);
+
+	disc_complete(client, err == 0);
+}
+
+static int cfg_active_preset_index_ntf(struct bt_has_client *client, uint16_t handle)
+{
+	BT_DBG("conn %p handle 0x%04x", client->conn, handle);
+
+	client->active_preset_subscription.notify = active_preset_idx_ntf_cb;
+	client->active_preset_subscription.write = cfg_active_preset_idx_ntf_cb;
+	client->active_preset_subscription.value_handle = handle;
+	client->active_preset_subscription.ccc_handle = 0x0000;
+	client->active_preset_subscription.end_handle = 0xffff;
+	client->active_preset_subscription.disc_params = &client->params.discover;
+	client->active_preset_subscription.value = BT_GATT_CCC_NOTIFY;
+	atomic_set_bit(client->active_preset_subscription.flags, BT_GATT_SUBSCRIBE_FLAG_VOLATILE);
+
+	return bt_gatt_subscribe(client->conn, &client->active_preset_subscription);
+}
+
+static uint8_t disc_active_preset_idx_cb(struct bt_conn *conn, uint8_t err,
+					 struct bt_gatt_read_params *read, const void *data,
+					 uint16_t length)
+{
+	struct bt_has_client *client = client_by_conn(conn);
+	struct net_buf_simple buf;
+
+	BT_DBG("conn %p err 0x%02x len %u", conn, err, length);
+
+	if (err || length == 0) {
+		disc_complete(client, false);
+		return BT_GATT_ITER_STOP;
+	}
+
+	BT_DBG("handle 0x%04x", read->by_uuid.start_handle);
+
+	net_buf_simple_init_with_data(&buf, (void *)data, length);
+
+	client->has.active_id = net_buf_simple_pull_u8(&buf);
+
+	if (cfg_active_preset_index_ntf(client, read->by_uuid.start_handle) < 0) {
+		disc_complete(client, false);
+	}
+
+	return BT_GATT_ITER_STOP;
+}
+
+static int disc_active_preset_idx(struct bt_has_client *client)
+{
+	BT_DBG("conn %p", client->conn);
+
+	client->params.read.func = disc_active_preset_idx_cb;
+	client->params.read.handle_count = 0u;
+	memcpy(&client->params.uuid, BT_UUID_HAS_ACTIVE_PRESET_INDEX, sizeof(client->params.uuid));
+	client->params.read.by_uuid.uuid = &client->params.uuid.uuid;
+	client->params.read.by_uuid.start_handle = 0x0001;
+	client->params.read.by_uuid.end_handle = 0xffff;
+
+	return bt_gatt_read(client->conn, &client->params.read);
```

Comment:
Why does this discovery function use `bt_gatt_read` while the discovery functions for the two other characteristics use `bt_gatt_discover`?

Reply:
Because it performs Read Using Characteristic UUID GATT operation. So doing 1 operation, the characteristic is read and discovered.